silk-core/src/main/kotlin/net/silkmc/silk/core/math/vector/VectorOperations.kt

```diff
@@ -45,17 +65,56 @@ operator fun Vector3f.component3() = z()
+
 operator fun Vec3i.times(n: Number): Vec3i = multiply(n.toInt())
 operator fun Vec3i.div(n: Number) = (1.0 / n.toDouble()).let { nDouble -> Vec3i(x * nDouble, y * nDouble, z * nDouble) }
-operator fun Vec3i.compareTo(n: Number) = Vec3(x.toDouble(), y.toDouble(), z.toDouble()).length().compareTo(n.toDouble())
+operator fun Vec3i.compareTo(n: Number) = Vec3(x, y, z).length().compareTo(n.toDouble())
 
 inline operator fun <reified T : Vec3i> T.plus(vec: Vec3i): T = offset(vec) as T
 inline operator fun <reified T : Vec3i> T.minus(vec: Vec3i): T = offset(!vec) as T
 operator fun Vec3i.times(vec: Vec3i): Vec3i = Vec3i(x * vec.x, y * vec.y, z * vec.z)
 operator fun Vec3i.div(vec: Vec3i): Vec3i = Vec3i(x / vec.x, y / vec.y, z / vec.z)
-operator fun Vec3i.compareTo(vec: Vec3i) = Vec3(x.toDouble(), y.toDouble(), z.toDouble()).lengthSqr().compareTo(Vec3(vec.x.toDouble(), vec.y.toDouble(), vec.z.toDouble()).lengthSqr())
+operator fun Vec3i.compareTo(vec: Vec3i) =
+    Vec3(x, y, z).lengthSqr().compareTo(Vec3(vec.x, vec.y, vec.z).lengthSqr())
+
+fun Vec3i.triple() = Triple(x, y, z)
+fun Vec3i(x: Number, y: Number, z: Number) = Vec3i(x.toDouble(), y.toDouble(), z.toDouble())
 
 operator fun Vec3i.component1() = x
 operator fun Vec3i.component2() = y
 operator fun Vec3i.component3() = z
+
+operator fun Vector3d.not(): Vector3d = times(-1f)
+
+operator fun Vector3d.plus(n: Number): Vector3d =
+    copy().apply { n.toDouble().let { set(it + x, it + y, it + z) } }
+
+operator fun Vector3d.minus(n: Number): Vector3d =
+    copy().apply { n.toDouble().let { set(x - it, y - it, z - it) } }
+
+operator fun Vector3d.times(n: Number): Vector3d = copy().apply { n.toDouble().let { times -> scale(times) } }
+operator fun Vector3d.div(n: Number): Vector3d = times(1.0 / n.toDouble())
+operator fun Vector3d.compareTo(n: Number) = Vec3(x, y, z).length().compareTo(n.toDouble())
+
+operator fun Vector3d.plus(vec: Vector3d): Vector3d = copy().apply { add(vec) }
+operator fun Vector3d.minus(vec: Vector3d): Vector3d = Vector3d(x - vec.x, y - vec.y, z - vec.z)
+operator fun Vector3d.times(vec: Vector3d): Vector3d = Vector3d(x * vec.x, y * vec.y, z * vec.z)
+operator fun Vector3d.div(vec: Vector3d): Vector3d = Vector3d(x / vec.x, y / vec.y, z / vec.z)
+operator fun Vector3d.compareTo(vec: Vector3d) =
+    Vec3(x, y, z).lengthSqr().compareTo(Vec3(vec.x, vec.y, vec.z).lengthSqr())
+
+fun Vector3d.copy() = Vector3d(x, y, z)
+fun Vector3d.triple() = Triple(x, y, z)
+fun Vector3d(x: Number, y: Number, z: Number) = Vector3d(x.toDouble(), y.toDouble(), z.toDouble())
+
+operator fun Vector3d.component1() = x
+operator fun Vector3d.component2() = y
+operator fun Vector3d.component3() = z
+
+
+fun <T : Number> Triple<T, T, T>.vec3() = Vec3(first, second, third)
+fun <T : Number> Triple<T, T, T>.vec3i() = Vec3i(first, second, third)
+fun <T : Number> Triple<T, T, T>.vector3f() = Vector3f(first, second, third)
+fun <T : Number> Triple<T, T, T>.vector3d() = Vector3d(first, second, third)
```

Comment:
the common project style would be to prefix these functions with `to`, e.g. `toVec3()`, same goes for all the other places where a "to" is missing in this file